load config dynamic/static

b01901143 opened the pull request:
Sample go app loading configuration yaml files.
Use -dynamic flag to choose from either loading to dynamic of static structure.

b01901143 committed:
load config dynamic/static

spiffxp commented on README.md:
````diff
@@ -1 +1,33 @@
+## Usage
+Build go app
+```
+# build the config module
+cd config
+go build
````
I think this is more than you need.  I was able to do the following
```console
$ go build
$ ./sample-go-app --help
Usage of ./sample-go-app:
  -config-path string
    	Path to config.yaml.
  -dynamic
    	Load config.yaml to dynamic structure.
```

b01901143 replied:
True.
Revised to use single-lined 'go build' or 'go install'

spiffxp commented on config/parse_config.go:
```diff
@@ -9,22 +9,59 @@ import (
+type configuration struct {
+	ApiVersion string 				`yaml:"apiVersion"`
+	Kind string 					`yaml:"kind"`
+	Metadata struct {
+		Name string 				`yaml:"name"`
+		Labels struct {
+			Name string 			`yaml:"name"`
+		}							`yaml:"labels"`
+
+	}								`yaml:"metadata"`
```
Avoid nested struct definitions unless there's a very good reason for them.  Instead, prefer defining structs at the top level, so that each struct is just a set of fields. e.g.

```go
type Labels struct {
  Name string `yaml:"name"`
}
type Metadata struct {
  Name string    `yaml:"name"`
  Labels labels  `yaml:"labels"`
}
type configuration {
  ApiVersion string `yaml:"apiVersion"`
  Kind string       `yaml:"kind"`
  Metadata Metadata `yaml:"metadata"`
 // etc...
}
```

b01901143 replied:
Done

spiffxp commented on main.go:
```diff
@@ -36,13 +37,11 @@ func main() {
 
-	d, err := yaml.Marshal(conf)
-	fmt.Println(string(d))
+	conf.PrintStruct()
```
this segfaults if there's an error loading the config
```console
$ ./sample-go-app
Invalid options: %v required flag --config-path was unset
Error loading config.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x11092f1]

goroutine 1 [running]:
main.main()
	/Users/spiffxp/w/b01901143/sample-go-app/main.go:45 +0x111
```

Instead, consider exiting early above
```go
	conf, err := config.LoadConfig(o.configPath, o.dynamic)
	if err != nil {
		fmt.Println("Error loading config.")
		os.Exit(1)
	}
```

(same goes if there are invalid options)

spiffxp replied:
a slightly friendlier alternative is to figure out how to print out help if the required flags aren't passed in

b01901143 replied:
Done.
Also added a printHelp() function to print the required flags.

spiffxp commented on main.go:
```diff
@@ -5,13 +5,13 @@ import (
 
 type options struct {
 	configPath    string
+	dynamic    	  bool
```
It looks like this code hasn't been formatted with `gofmt`.  This would fail tests in a repo that ensures gofmt is used

You can do this manually by running `gofmt -w .` which will format the code in place

Alternatively, the easier way to do this is use an IDE or text editor that will automatically run gofmt on save

b01901143 replied:
Done.
Also set up gofmt package for my sublime editor.

spiffxp commented on config/parse_config.go:
```diff
@@ -9,22 +9,59 @@ import (
+}
+
+func (conf dynamicStruct) PrintStruct() {
+	fmt.Println(conf)
```
FWIW you may be able to Marshal this as wel

b01901143 replied:
Fixed, marshaled with yaml.Marshal().

spiffxp reviewed (changes requested):
This is mostly good.  The one thing lacking that I would expect of a typical go project is unit tests.  The thing is, there's not a lot to unit test here.  It would probably be better to get some practice writing unit tests against an existing codebase

b01901143 committed:
avoid nested struct

added printHelp() function for failed input flags
marshaled dynamic struct (map[interface{}]interface{}) for printing

spiffxp commented on config/parse_config.go:
```diff
@@ -1,30 +1,73 @@
+}
+
+// explicitly define the tags for each field
+type configuration struct {
```
Just a side note. I would typically expect data structs to be exported by default. I can't find a citable rule off-hand, but in general I find go tends to err on the side of data visibility, rather than data encapsulation.

spiffxp reviewed (approved):
LGTM
